src/coreclr/jit/lower.cpp

```diff
@@ -7934,6 +7941,97 @@ static bool GetStoreCoalescingData(Compiler* comp, GenTreeStoreInd* ind, StoreCo
     return true;
 }
 
+//------------------------------------------------------------------------
+// GetLoadStoreCoalescingData: perform legality check for store coalescing.
+//
+// Arguments:
+//    data1 - the data needed for store coalescing for the first load
+//    data2 - the data needed for store coalescing for the second node
+//
+// Return Value:
+//    true if two stores or two loads can be coalesced, false otherwise.
+//
+static bool CanBeCoalesced(const LoadStoreCoalescingData& data1, const LoadStoreCoalescingData& data2)
+{
+    // BaseAddr, Index, Scale and Type all have to match.
+    if ((data1.scale != data2.scale) || (data1.targetType != data2.targetType) ||
+        !GenTree::Compare(data1.baseAddr, data2.baseAddr) || !GenTree::Compare(data1.index, data2.index))
+    {
+        return false;
+    }
+
+    // Make sure both are either loads or stores and not a mix.
+    if ((data1.value == nullptr) != (data2.value == nullptr))
+    {
+        return false;
+    }
+
+    // Value should be either a constant or an IND in both cases.
+    if ((data1.value != nullptr) && !data1.value->OperIs(data2.value->OperGet()))
+    {
+        return false;
+    }
+
+    // Offset is the same - we can just remove the previous store.
+    if (data1.offset == data2.offset)
+    {
+        return true;
+    }
+
+    // Otherwise, the difference between two offsets has to match the size of the type or be zero
+    if (abs(data1.offset - data2.offset) != (int)genTypeSize(data1.targetType))
+    {
+        return false;
+    }
+
+    // Now the hardest part: decide whether it's safe to use an unaligned write.
+    //
+    // IND<byte> is always fine (and all IND<X> created here from such)
+    // IND<simd> is not required to be atomic per our Memory Model
+    const bool allowsNonAtomic = data1.allowsNonAtomic && data2.allowsNonAtomic;
```

Comment:
Would it make a difference (cover more cases) if it was a positive check: `allosAtomic`?

Reply:
I am not sure I follow - can you clarify?

Reply:
Was just wondering if the condition below actually care for "if atomics are allowed, do the transformation" instead of "if non-atomics are allowed, skip the transformation", then maybe it can be based on `allowAtomic`? Not sure if it's feasible or would make any difference. 😅

Reply:
The idea that we assume that we can't use wider loads by default, we need to find a proof that we can and we only have a few hints for that today so it's quite conservative